addons/mail/models/mail_template.py

```diff
@@ -144,6 +145,81 @@ def _search_template_category(self, operator, value):
 
         raise NotImplementedError(_('Operation not supported'))
 
+    def _validate_reference(self, variable, expression):
+        # Ignore context (ctx) while checking
+        if variable == "ctx":
+            return True
+        refs = expression.strip(' (){}').split('.')
+
+        VARIABLE_DICT = {
+            'user': self.env.user,
+            'env': self.env,
+            'object': self.env[self.model_id.model],
+        }
+
+        curr = VARIABLE_DICT[variable]
+        i = 0
+        if not isinstance(refs[i], str):
+            raise ValidationError(_("Invalid reference used!"))
+        while i < len(refs):
+            if refs[i] in dir(curr):
+                if isinstance(curr, models.BaseModel) and refs[i] in curr._fields:
+                    if i < len(refs) - 1 and not curr._fields[refs[i]].relational:
+                        raise ValidationError(_("Invalid Field: %(field)s in %(model)s is not relational", field=refs[i], model=curr))
+                    curr = curr[refs[i]]
+                elif refs[i] in VARIABLE_DICT:
+                    curr = VARIABLE_DICT[refs[i]]
+                else:
+                    break
+            else:
+                raise ValidationError(_("Invalid Attribute: %(attribute)s not in %(model)s", attribute=refs[i], model=curr))
+            i += 1
+        return True
+
+    def _parse_expression(self, expression):
+        if not expression:
+            return
+
+        try:
+            expressions = re.findall(r"\{\{.*?\}\}", expression)
+            # Check if user is allowed when dynamic field is added
+            if expressions and not (self.env.user.has_group('mail.group_mail_template_editor') or self.env.su):
+                raise AccessError(_("Access Denied!"))
+            for e in expressions:
+                parsed_expr = self.env['ir.qweb']._compile_expr(e)
+                lst = re.findall(r"values\['(.+?)'\]\.([\w\._]+)", parsed_expr)
+                for var, expr in lst:
+                    self._validate_reference(var, expr)
+        except SyntaxError as e:
+            raise ValidationError(_("%s", str(e)))
+
+        return True
+
+    def _validate_template(self):
+        self.ensure_one()
+
+        mail_template_fields = [
+            'subject',
+            'email_from',
+            'email_to',
+            'partner_to',
+            'email_cc',
+            'reply_to',
+        ]
+
+        for field in mail_template_fields:
+            self._parse_expression(self[field])
+
+    @api.constrains('body_html', 'model_id', 'subject', 'email_from', 'email_to', 'partner_to', 'email_cc', 'reply_to')
```

Comment:
I think it's a little too heavy for the constraint to check. Can we try something in create and write that way we can check only modified field